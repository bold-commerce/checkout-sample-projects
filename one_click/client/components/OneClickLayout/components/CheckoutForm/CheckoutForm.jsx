import React  from 'react';
import { Route, useLocation, Switch } from "react-router-dom";
import { TransitionGroup, CSSTransition } from "react-transition-group";
import { BillingAddress } from '@boldcommerce/checkout-react-components';
import { Shipping } from '../Shipping';

const CheckoutForm = () => {
  const location = useLocation();
  return (
    <>
      <TransitionGroup>
        <CSSTransition
          timeout={150}
          classNames="fade"
          key={location.key}
        >
          <Switch location={location}>
            <Route exact path="/shipping" component={Shipping} />
            <Route exact path="/billing" component={BillingAddress} />
          </Switch>
        </CSSTransition>
      </TransitionGroup>
    </>
  )};


export default CheckoutForm;
