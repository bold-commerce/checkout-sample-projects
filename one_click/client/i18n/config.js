import i18n from 'i18next';
import enTranslation from './en/translation.json';
import { initReactI18next } from 'react-i18next';

i18n
  .use(initReactI18next)
  .init({
    resources: {
      en: {
        translation: enTranslation
      }
    },
    lng: 'en',
    fallbackLng: 'en',
    interpolation: {
      escapeValue: false, 
    }
  });

export default i18n;