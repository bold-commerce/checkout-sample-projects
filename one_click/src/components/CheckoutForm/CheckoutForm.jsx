import React, { useEffect, useRef, useState }  from 'react';
import { Route, Routes } from "react-router-dom";
import { useLocation, useNavigate } from 'react-router';
import { useCheckoutStore, useCustomer, useLineItems } from '@boldcommerce/checkout-react-components';
import { Summary } from '../Summary';
import { Shipping } from '../Shipping';
import { Billing } from '../Billing';
import { useErrorLogging, useInventory } from '../../hooks';
import { Inventory } from '../../pages/Inventory';
import { ProcessingOrder } from '../../pages/Processing';
import { Confirmation } from '../../pages/Confirmation';
import { CheckoutButton } from '../CheckoutButton';
import classNames from 'classnames';
import { LoadingState } from '../LoadingState';
import { IndexPageGuest, IndexPage } from '../../pages/IndexPage';

const CheckoutForm = () => {
  const { data: lineItems } = useLineItems();
  const { state } = useCheckoutStore();
  const { data: customer } = useCustomer();
  const orderStatus = state.orderInfo.orderStatus;
  const location = useLocation();
  const checkInventory = useInventory();
  const logError = useErrorLogging();
  const navigate = useNavigate();
  const [openSection, setOpenSection] = useState('/');
  const [loading, setLoading] = React.useState(true); // calling React.useState to be able to mock with snapshot tests.
  const [height, setHeight] = useState(null);
  const mainEl = useRef(null);
  const summaryEl = useRef(null);
  const shippingEl = useRef(null);
  const billingEl = useRef(null);
  const processingEl = useRef(null);
  const confirmationEl = useRef(null);
  const inventoryEl = useRef(null);
  const checkoutButtonEl = useRef(null);

  const showCheckoutButton = openSection === 'summary' || (openSection === '/' && customer.platform_id);
  const renderSidebar = openSection === '/' || openSection.indexOf('/') === -1;
  const getInventory = async () => {
    try{
      const inventory = await checkInventory(lineItems);
      if (inventory) {
        navigate('/inventory', inventory)
      }
    } catch (e) {
      logError('check_inventory', e);
    }
  }
  useEffect(() => {
    setOpenSection(location.pathname);
  }, [location.pathname]);

  useEffect(async () => {
    setLoading(true);
    await getInventory();
    setLoading(false);
  }, []);

  useEffect(() => {
    if (orderStatus === 'processing') {
      navigate('/processing');
    } else if (orderStatus === 'completed') {
      navigate(`/confirmation?public_order_id=${state.publicOrderId}`);
    }
  }, [orderStatus]);

  useEffect(() => {
    let openEl = mainEl;
    switch(openSection){
      case "summary":
      case "summaryB":
        openEl = summaryEl;
        break;
      case "shipping":
        openEl = shippingEl;
        break;
      case "billing":
        openEl = billingEl;
        break;
      case "/processing":
        openEl = processingEl;
        break;
      case "/confirmation":
        openEl = confirmationEl;
        break;
      case "/inventory":
        openEl = inventoryEl;
        break;
    }

    const resizeObserver = new ResizeObserver(() => {  
      if(showCheckoutButton){    
        setHeight(openEl.current?.clientHeight + checkoutButtonEl.current?.clientHeight);
      }
      else {
        setHeight(openEl.current?.clientHeight);
      }
    });

    if (openEl.current) {
      resizeObserver.observe(openEl.current);
    }
  
    return () => {
      resizeObserver.disconnect();
    }
  }, [openSection, mainEl?.current]);

  const style = window.innerWidth > 768 ? {
    height: height? `${height}px` : null,
    overflowY: height > window.innerHeight ? 'scroll' : 'hidden',
  } : {}

  return (
    <div className="Checkout__Form" style={style}>
      { (loading && <LoadingState/>) || 
        <>
          <Routes location={location}>
            <Route exact path="/processing" element={<ProcessingOrder ref={processingEl} />} />
            <Route exact path="/confirmation" element={<Confirmation ref={confirmationEl}/>} />
            <Route exact path="/inventory" element={<Inventory ref={inventoryEl}/>} />
            {
              customer.platform_id ?
              <Route exact path="/" element={<IndexPage ref={mainEl} onSectionChange={setOpenSection} show={openSection==="/"}/>} /> :
              <Route exact path="/" element={<IndexPageGuest ref={mainEl} onSectionChange={setOpenSection} show={openSection==="/"}/>} />
            }
          </Routes>
          { renderSidebar ? 
            <>
              <Shipping ref={shippingEl} show={openSection==='shipping'} onBack={() => setOpenSection("/")} />
              <Summary ref={summaryEl} section={openSection} onSectionChange={setOpenSection}/>
              <Billing ref={billingEl} section={openSection} onSectionChange={setOpenSection}/>
            </>
            : null
          }
          <CheckoutButton ref={checkoutButtonEl} className={(classNames("CheckoutButton", "CheckoutButton__Desktop", showCheckoutButton ? "CheckoutButton__Show" : "CheckoutButton__Hide"))} />
        </>
      }
    </div>
  )};


export default CheckoutForm;
