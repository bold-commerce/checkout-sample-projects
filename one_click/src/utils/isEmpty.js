const isEmpty = (obj) => {
  return JSON.stringify(obj) === JSON.stringify({});
};

export default isEmpty;