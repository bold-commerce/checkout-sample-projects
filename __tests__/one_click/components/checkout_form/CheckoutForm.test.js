import React from "react";
import CheckoutForm from '../../../../one_click/src/components/CheckoutForm/CheckoutForm'
import ResizeObserver from "../../../../__mocks__/ResizeObserver";
import { render, waitFor } from '@testing-library/react';
import { exampleLineItems as MOCKexampleLineItems } from '../../../utils/lineItemHelpers';
import { testApplicationState as MOCKapplicationState } from '../../../utils/applicationStateHelper';
import { exampleShippingState as MOCKexampleShippingState } from '../../../utils/shippingLinesHelper';
import {
    exampleAddress as MOCKexampleAddress,
    countries as MOCKcountries,
    caProvinces as MOCKprovinces,
    exampleSavedAddresses as MOCKsavedAddresses
} from '../../../utils/addressHelpers';
import '../../../../one_click/src/i18n/config'

jest.mock('@boldcommerce/checkout-react-components', () => ({
    useCheckoutStore: () => ({
        state: {
            orderInfo: {
                billingSameAsShipping:true,
                orderStatus: ""
            },
            errors: { order: null },
            loadingStatus: { isLoading: false },
            applicationState: MOCKapplicationState,
            orderTotals: {
                taxesTotal: 1200,
                subTotal: 23000,
                total: 24200
            }
        }
    }),
    useLineItems: () => ({
        data: MOCKexampleLineItems
    }),
    useShippingLines: () => ({
        data: MOCKexampleShippingState
    }),
    useShippingAddress: () => ({
        data: MOCKexampleAddress,
        submitShippingAddress: (() => {})
    }),
    useCustomer: () => ({
        data: {
            platform_id: 123
        }
    }),
    useBillingAddress: () => ({
        data: MOCKexampleAddress
    }),
    usePaymentIframe: () => ({
        data: {
            url: 'test.url',
            loadingStatus: 'fulfilled',
            paymentIframeOnLoaded: (() => {})
        }
    }),
    useDiscount: () => ({
        data: {
            discountTotal: 1234,
            discountApplied: true,
            discountCode: '?'
        },
        applyDiscount: (() => {}),
        removeDiscount: (() => {})
    }),
    useBillingSameAsShipping: () => ({
        data: true,
        setBillingSameAsShipping: (() => {})
    }),
    useCountryInfo: () => ({
        data: {
            countries: MOCKcountries,
            provinceLabel: "province",
            provinces: MOCKprovinces,
            showPostalCode: true,
            showProvince: true
        }
    }),
    useSavedAddresses: () => ({
        data: MOCKsavedAddresses
    }),
    useLoadingStatus: () => ({
        shippingAddressLoadingStatus: 'complete',
        shippingLinesLoadingStatus: 'complete'
    })
})).mock('react', () => ({
    ...jest.requireActual('react'),
    useContext: () => ({
        websiteName: 'TestSite'
    }),
})).mock('react-router', () => ({
    ...jest.requireActual('react-router'),
    useLocation: () => ({
        pathname: '/' 
    }),
    useNavigate: () => ({
        navigate: (() => {})
    }),
    Routes: () => ({
        useRoutes: () => {
        <route 
        key='key'
        name='screen name'
        path='/'
        />
    }})
})).mock('../../../../one_click/src/hooks', () => ({
    ...jest.requireActual('../../../../one_click/src/hooks'),
    useInventory: () => ({
        checkInventory: (() => {})
    })
}));

describe('CheckoutForm', () => {
    test('renders CheckoutForm', async () => {
        // React.useState = jest.fn().mockReturnValue([false, (i) => {}])

        const { asFragment } = render(
            <CheckoutForm />
        );

        await waitFor(() => {
            expect(asFragment()).toMatchSnapshot();
        });
    });
});
