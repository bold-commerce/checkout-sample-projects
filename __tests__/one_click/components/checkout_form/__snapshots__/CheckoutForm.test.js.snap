// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CheckoutForm renders CheckoutForm 1`] = `
<DocumentFragment>
  <div
    class="Checkout__Form"
    style="overflow-y: hidden;"
  >
    <div
      class="FieldSet--LoadingState"
    >
      <span
        class="stx-loading-spinner"
      />
    </div>
  </div>
</DocumentFragment>
`;
